lms/services/http.py

```diff
@@ -0,0 +1,161 @@
+
+            This can be a (user, pass) 2-tuple which will be serialized using
+            HTTP Basic Authentication.
+
+            Or it can be a callable in order to implement other authentication
+            schemes. `requests` itself and libraries like `requests_oauthlib`
+            provide auth callables that you can use or you can write your own.
+
+            See: https://docs.python-requests.org/en/master/user/authentication/
+
+        :param timeout: How long (in seconds) to wait before raising an error.
+
+            This can be a (connect_timeout, read_timeout) 2-tuple or it can be
+            a single float that will be used as both the connect and read
+            timeout.
+
+            Note that the read_timeout is *not* a time limit on the entire
+            response download. It's a time limit on how long to wait *between
+            bytes from the server*. The entire download can take much longer.
+
+        :param schema: A schema class to use to validate the response
+        :type schema: lms.validation.RequestsResponseSchema
+
+        :param kwargs: Any other keyword arguments will be passed directly to
+            requests.Session().request():
+            https://docs.python-requests.org/en/latest/api/#requests.Session.request
+
+        :raise HTTPError: If sending the request or receiving the response
+            fails (DNS failure, refused connection, timeout, too many
+            redirects, etc).
+
+            The original exception from requests will be available as
+            HTTPError.__cause__.
+
+            In this case HTTPError.response will be None.
+
+        :raise HTTPError: If an error response (4xx or 5xx) is received.
+
+            The original exception from requests will be available as
+            HTTPError.__cause__.
+
+            The error response will be available as HTTPError.response.
+
+        :raise HTTPValidationError: If a `schema` argument is given and the
+            response fails validation with the schema.
+
+            The original lms.validation.ValidationError will be available as
+            HTTPValidationError.__cause__.
+
+            The invalid response will be available as
+            HTTPValidationError.response.
+        """
+        response = None
+
+        try:
+            response = self._session.request(
+                method,
+                url,
+                params=params,
+                data=data,
+                json=json,
+                headers=headers,
+                auth=auth,
+                timeout=timeout,
+                **kwargs,
+            )
+            response.raise_for_status()
+        except RequestException as err:
+            raise HTTPError(response) from err
+
+        if schema:
+            try:
+                response.validated_data = schema(response).parse()
+            except ValidationError as err:
+                raise HTTPValidationError(response) from err
+
+        return response
+
```

Comment:
Would adding `post` and `get` utility functions here make sense?

Doc string could point to the docs on the main request func and the actual function bodies be pretty minimal

https://github.com/psf/requests/blob/master/requests/api.py#L64

Reply:
Yeah, I think it could be easily added

Reply:
Added these: https://github.com/hypothesis/lms/pull/2711/commits/7ca7666d4e3e41cee9201b2bd0838b444a9c4610

It's actually possible to do this in a fancy way in `HTTPService.__init__()`:

```python
for http_verb in ("get", "put", "post", "patch", "delete"):
    setattr(self, http_verb, partial(self.request, http_verb))
```

But I don't think it's worth it: it just means that `mock` fails to detect the `get()`, `put()` etc methods and you then have to work around that.